bot/cogs/free.py

```diff
@@ -0,0 +1,90 @@
+import logging
+from datetime import datetime
+
+from discord import Colour, Embed, Member, utils
+from discord.ext.commands import BucketType, Context, command, cooldown
+
+
+log = logging.getLogger(__name__)
+
+
+class Free:
+    """Tries to figure out which help channels are free."""
+
+    PYTHON_HELP_ID = 356013061213126657
+    TIME_INACTIVE = 300
+
+    @command(name="free", aliases=('f',))
+    @cooldown(1, 60.0, BucketType.channel)
+    async def free(self, ctx: Context, user: Member = None, seek: int = 2):
+        """
+        Lists free help channels by likeliness of availability.
+        :param user: accepts user mention, ID, etc.
+        :param seek: How far back to check the last active message.
+
+        seek is used only when this command is invoked in a help channel.
+
+        When seek is 2, we are avoiding considering the last active message
+        in a channel to be the one that invoked this command.
+
+        When seek is 3 or more, a user has been mentioned on the assumption
+        that they asked if the channel is free or they asked their question
+        in an active channel, and we want the message before that happened.
+        """
+        free_channels = []
+        python_help = utils.get(ctx.guild.categories, id=self.PYTHON_HELP_ID)
+
+        if user is not None and seek == 2:
+            seek = 3
+        elif not 0 < seek < 10:
+            seek = 3
+
+        for channel in python_help.channels:
+            if channel.id == ctx.channel.id:
+                messages = await channel.history(limit=seek).flatten()
+                msg = messages[seek-1]
+            else:
+                messages = await channel.history(limit=1).flatten()
+                msg = messages[0]
```

Comment:
This line and line 45 will throw this exception:

```py
Traceback (most recent call last):
  File "/home/zappa/.local/share/virtualenvs/bot-05f_83Kn/src/discord-py/discord/client.py", line 225, in _run_event
    await coro(*args, **kwargs)
  File "/home/zappa/pydis/bot/bot/cogs/events.py", line 146, in on_command_error
    raise e.original
  File "/home/zappa/.local/share/virtualenvs/bot-05f_83Kn/src/discord-py/discord/ext/commands/core.py", line 61, in wrapped
    ret = await coro(*args, **kwargs)
  File "/home/zappa/pydis/bot/bot/cogs/free.py", line 48, in free
    msg = messages[0]
IndexError: list index out of range
```
If there are not enough messages in the channel.

This may not be an issue now (there are plenty of messages in them), but cleaning/creating help-channels on the fly has been discussed in the past and implementing that could break this. Don't know if it's relevant, though.

Reply:
Yep I knew someone would point this out. I think it's probably a good idea to implement this just for the sake of completeness. I too have doubts there will ever be a help channel that's empty, but doesn't mean it's not worth doing!

Reply:
There will never be an empty help channel. If we switch a new system where we create help channels on the fly, this command will become irrelevant, and will be removed. So I'm gonna resolve this. I'm all for solving edge cases, but this isn't on the edge, it's off the surface entirely.